app/models/podcast_planner.rb

```diff
@@ -0,0 +1,152 @@
+class PodcastPlanner
+  attr_accessor :podcast_id, :dates, :selected_days, :week_condition, :period, :monthly_weeks, :start_date, :date_range_condition, :number_of_episodes, :end_date, :publish_time, :segment_count, :drafts
+
+  def initialize(params = {})
+    @dates = []
+    @drafts = []
+    @podcast_id = params[:podcast_id]
+    @start_date = params[:start_date].try(:to_datetime)
+    @selected_days = params[:selected_days].try { reject(&:blank?) }.try { map { |day| day.to_i } }
+    @period = params[:period].try(:to_i)
+    @monthly_weeks = params[:monthly_weeks].try { reject(&:blank?) }.try { map { |week| week.to_i } }
+    @date_range_condition = params[:date_range_condition]
+    @week_condition = params[:week_condition]
+    @number_of_episodes = params[:number_of_episodes].try(:to_i)
+    @end_date = params[:end_date].try(:to_datetime)
+    @publish_time = params[:publish_time].try(:to_time)
+    @segment_count = params[:segment_count].try(:to_i)
+  end
+
+  def date_range_ends_by_episodes?
+    @date_range_condition == "episodes"
+  end
+
+  def date_range_ends_by_end_date?
+    @date_range_condition == "date"
+  end
+
+  def episodes_remain?
+    @dates.length < @number_of_episodes
+  end
+
+  def end_date_passed?(date)
+    date > @end_date
+  end
+
+  def periodic?
+    @week_condition == "periodic"
+  end
+
+  def monthly?
+    @week_condition == "monthly"
+  end
+
+  def ready_to_select_weeks?
+    if periodic?
+      @period.present?
+    elsif monthly?
+      @monthly_weeks.present?
+    else
+      false
+    end
+  end
+
+  def ready_to_select_date_range?
+    if date_range_ends_by_episodes?
+      @number_of_episodes.present?
+    elsif date_range_ends_by_end_date?
+      @end_date.present? && @end_date > @start_date
+    else
+      false
+    end
+  end
+
+  def ready_to_generate_dates?
+    @selected_days.present? &&
+      ready_to_select_weeks? &&
+      ready_to_select_date_range?
+  end
+
+  def generate_dates!
+    return unless ready_to_generate_dates?
+
+    if date_range_ends_by_episodes?
+      generate_dates_by_remaining_episodes
+    elsif date_range_ends_by_end_date?
+      generate_dates_by_end_date
+    end
+  end
+
+  def clear_dates!
+    @dates = []
+  end
+
+  def select_days_of_week_to_add(current_day)
+    dates_to_add = []
+    7.times do |day|
+      dates_to_add.push(current_day) if @selected_days.include?(current_day.wday)
+      current_day += 1.day
+    end
+    dates_to_add
+  end
+
+  def generate_dates_by_remaining_episodes
+    current_day = @start_date
+
+    while episodes_remain?
```

Comment:
the logic inside of this loop is the exact same as in the method below, which means i should be able to extract it. (something like, `add_dates_and_advance_date_pointer` or something?). the conditional bothers me, because the two branches are just different enough to require two different methods, and i'm wondering if adding so many different methods pollutes the logic clarity (or does it help)? is there another way to do this?

Reply:
I think this is fairly readable.  If you're interested in a slightly different implementation, you could look at the [Publish code](https://github.com/PRX/publish.prx.org/blob/main/src/app/series/directives/series-plan.component.ts#L109).

Kind of similar to a while.  But the main difference is I had it walk through the entire calendar one day at a time, rather than adding weeks to it.  And for each day, it checks:

- Are we past max episodes? Exit early.
- Are we past the max date? Exit early.
- Is the date one of the days-of-week we should add?  And is it one of the weeks-of-month we should add?  Then add it.